src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -0,0 +1,379 @@
+        level_handlers[target_level].add_pending_task(next_task_id, &target_level_inputs);
+        // Here, we have known that `select_level_input` is valid
+        let mut splits = Vec::with_capacity(target_level_inputs.len());
+        splits.push(KeyRange::new(Bytes::new(), Bytes::new()));
+        if target_level_inputs.len() > 1 {
+            for table in &target_level_inputs[1..] {
+                let key_before_last: Bytes = FullKey::from_user_key_slice(
+                    user_key(&table.key_range.as_ref().unwrap().left),
+                    HummockEpoch::MAX,
+                )
+                .into_inner()
+                .into();
+                splits.last_mut().unwrap().right = key_before_last.clone();
+                splits.push(KeyRange::new(key_before_last, Bytes::new()));
+            }
+        }
+
+        Some(SearchResult {
+            select_level: Level {
+                level_idx: select_level as u32,
+                level_type: LevelType::Overlapping as i32,
+                table_infos: select_level_inputs,
+            },
+            target_level: Level {
+                level_idx: target_level as u32,
+                level_type: LevelType::Nonoverlapping as i32,
+                table_infos: target_level_inputs,
+            },
+            split_ranges: splits,
+        })
+    }
+
+    fn pick_target_level_overlap_files(
+        &self,
+        tier_key_range: &KeyRange,
+        level: &Level,
+        level_handlers: &LevelHandler,
+        input_ssts: &mut Vec<SstableInfo>,
+    ) -> bool {
+        let overlap_begin = level.table_infos.partition_point(|table_status| {
+            user_key(&table_status.key_range.as_ref().unwrap().right)
+                < user_key(&tier_key_range.left)
+        });
+        if overlap_begin >= level.table_infos.len() {
+            return true;
+        }
+        // pick up files in L1 which are overlap with L0 to target level input.
+        for table in &level.table_infos[overlap_begin..] {
+            let end_range = KeyRange::from(table.key_range.as_ref().unwrap());
+            if user_key(&end_range.left) > user_key(&tier_key_range.right) {
+                break;
+            }
+            if level_handlers.is_pending_compact(&table.id) {
+                return false;
+            }
+            input_ssts.push(table.clone());
+        }
+        true
+    }
+
+    fn pick_intra_l0_compaction(
+        &self,
+        level0: &Level,
+        level0_handler: &mut LevelHandler,
+    ) -> Option<SearchResult> {
+        if level0.table_infos.len() < self.level0_max_file_number {
+            return None;
+        }
+
+        let mut sst_idx = 0;
+        while sst_idx < level0.table_infos.len() {
+            if !level0_handler.is_pending_compact(&level0.table_infos[sst_idx].id) {
+                break;
+            }
+            sst_idx += 1;
+        }
+
+        if level0.table_infos.len() - sst_idx < self.level0_trigger_number {
```

Comment:
Use <= in case level0_trigger_number is set to 0 in some cases?

Reply:
🤔could level0_trigger_number be 0?